### PR TITLE
Update floor_divide behavior in line with NumPy 1.20

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50438 Support JIT shape analysis for div with rounding_mode argument
* #55225 grad_fn getter for optional strings
* #51989 Sparse support for division rounding_mode argument
* **#56893 Update floor_divide behavior in line with NumPy 1.20**
* #55380 Vectorize floating point floor_divide

Fixes gh-56814

Differential Revision: [D28025814](https://our.internmc.facebook.com/intern/diff/D28025814)

### `torch.divide` with `rounding_mode='floor'` now returns infinity when a non-zero number is divided by zero

This fixes the `rounding_mode='floor'` behaviour to return the same non-finite values as other rounding modes when there is a division by zero. Previously it would always result in a NaN value, but a non-zero number divided by zero should return +/- infinity in IEEE floating point arithmetic.

Note this does not effect `torch.floor_divide` or the floor division operator, which currently use `rounding_mode='trunc'` (and are also deprecated for that reason).

Example:

Version 1.8.1:
```
>>> a = torch.tensor([-1.0, 0.0, 1.0])
>>> b = torch.tensor([0.0])
>>> torch.divide(a, b, rounding_mode='floor')
tensor([nan, nan, nan])
```

Version 1.9.0:
```
>>> a = torch.tensor([-1.0, 0.0, 1.0])
>>> b = torch.tensor([0.0])
>>> torch.divide(a, b, rounding_mode='floor')
tensor([-inf, nan, inf])
```

### CODE REVIEW

File: test/test_binary_ufuncs.py [hunk 1]
```diff
@@ -335,15 +337,38 @@ def test_div_rounding_nonfinite(self, device, dtype):
                              exact_device=False, exact_dtype=exact_dtype)
 
         # Compare contiguous (likely vectorized) against non-contiguous (not vectorized)
-        storage = torch.empty((20, 20), dtype=dtype, device=device)
-        storage[::2, ::2] = a
-        storage[1::2, 1::2] = b
+        a_noncontig = torch.empty([2 * i for i in a.shape], dtype=dtype, device=device)[::2, ::2]
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:340]: Do we really want empty tensors here?

Also make_tensor now has a `noncontiguous` kwarg to create noncontiguous tensors

File: test/test_binary_ufuncs.py [hunk 2]
```diff
@@ -335,15 +337,38 @@ def test_div_rounding_nonfinite(self, device, dtype):
...
+                         dtype=dtype)
+        exact_dtype = (dtype != torch.bfloat16)
+        if exact_dtype:
+            an = a.cpu().numpy()
+        else:
+            an = a.float().cpu().numpy()
+
+        zero = torch.zeros_like(a)
+
+        # NOTE: NumPy's floor_divide rounding changed in 1.20.0 to be consistent with divide
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:362]: Does this test need to check for the NumPy version being at least 1.2?

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -355,9 +380,9 @@ def test_div_rounding_numpy(self, device, dtype):
         a = make_tensor((4096,), device, dtype, low=low, high=high)
         b = make_tensor((4096,), device, dtype, low=low, high=high)
 
-        # Avoid integer division by zero which raises
-        if not dtype.is_floating_point:
-            b[b == 0] = 1
+        # Avoid division by zero which raises for integers and, for floats,
+        # NumPy behavior changed in 1.20
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:384]: What should this updated comment be given the change in behavior?

**Comment by mruberry** [comment]
Hey @peterbell! This looks great; I just made a few suggestions

Is it worth updating our docs for this behavior, too?

Are there are other functions that need a similar fix for consistency?

File: test/test_binary_ufuncs.py (hunk 1 above)
Comment (peterbell10) [reply test/test_binary_ufuncs.py:340]: > Do we really want empty tensors here?

Yes, it gets filled on the next line.

> Also make_tensor now has a noncontiguous kwarg to create noncontiguous tensors

I don't want random data, I specifically want `a` and `b` copied into a non-contiguous tensor.

File: test/test_binary_ufuncs.py (hunk 2 above)
Comment (peterbell10) [reply test/test_binary_ufuncs.py:362]: Nope, `np.divide` hasn't changed.

**Comment by peterbell10** [comment]
> Is it worth updating our docs for this behavior, too?

The docs don't actually specify what happens when you divide by zero. 

> Are there are other functions that need a similar fix for consistency?

Not that I'm aware of. The counterpart to `floor_divide` is `remainder` which seems to return the same result as before for division by zero.

File: test/test_binary_ufuncs.py (hunk 3 above)
Comment (peterbell10) [reply test/test_binary_ufuncs.py]: @mruberry I've added that they changed to comply with IEEE rules for inf/nan.

**Review by mruberry — APPROVED** [review]
Cool! Thanks for the update, @peterbell10!

**Pull request closed without merging**